docs/jep/14273-shard-map.md

```diff
@@ -0,0 +1,640 @@
+backward pass, exploiting linearity. Many advanced `pmap` users addressed this
+challenge by using `custom_vjp` to implement `psum_idrev` and `id_psumrev`
+functions, but since it's easy to accidentally leave those imbalanced, that
+technique is a foot-cannon. We have some ideas on how to provide this
+functionality in a safer way.
+
+## When should you use `shmap` and when should you use `pjit`?
+
+One philosophy is: it is almost always simpler to write a program in `jit==pjit`
+&mdash; but if a given part of the program is less optimized by the compiler than it
+could be, drop into `shmap`!
+
+### A realistic transformer example
+
+In fact, we can implement a simple version of the ["collective
+matmul"](https://dl.acm.org/doi/pdf/10.1145/3567955.3567959) algorithm
+recently introduced in XLA to overlap communication and computation using `shmap`
+and 30 lines of Python. The basic idea of the algorithm can be grasped with a
+simple example.
+
+Suppose we want to compute `C = A @ B` where `A` is sharded by a 1D mesh on the
+0-th dimension while `B` and `C` are replicated.
+
+```python
+M, K, N = 4096, 2048, 1024
+A = jnp.arange(np.prod((M, K))).reshape((M, K))
+B = jnp.arange(np.prod((K, N))).reshape((K, N))
+
+mesh = Mesh(np.array(jax.devices()), axis_names=('x'))
+A_x = jax.device_put(A, NamedSharding(mesh, P('x', None)))
+
+@jax.jit
+def f(lhs, rhs):
+  return lhs @ rhs
+
+C = f(A_x, B)
+```
+
+A profile shows the blocking all-gather across 8 devices before the matmul can
+start. This is suboptimal because `A` is sharded on a non-contracting dimension,
+and each shard of `A` can be matmul'ed with `B` independently and this chunked
+computation can be overlapped with fetching of the next shard of `A` from
+another device.
+
+<img width="1147" alt="image" src="https://user-images.githubusercontent.com/1458824/216507011-e854fb11-43d5-484d-993b-19a3349ed4b9.png">
+
+This overlap can be implemented using `shmap` and explicit collectives.
+
+```python
+def collective_matmul_allgather_lhs_non_contracting(lhs, rhs):
+  # lhs is the looped operand; rhs is the local operand
+  axis_size = jax.lax.psum(1, axis_name='x')
+  axis_index = jax.lax.axis_index(axis_name='x')
+  chunk_size = lhs.shape[0]
+
+  def f(i, carrys):
+    accum, lhs = carrys
+    # matmul for a chunk
+    update = lhs @ rhs
+    # circular shift to the left
+    lhs = jax.lax.ppermute(
+        lhs,
+        axis_name='x',
+        perm=[(j, (j - 1) % axis_size) for j in range(axis_size)]
+    )
+    # device 0 computes chunks 0, 1, ...
+    # device 1 computes chunks 1, 2, ...
+    update_index = (((axis_index + i) % axis_size) * chunk_size, 0)
+    accum = jax.lax.dynamic_update_slice(accum, update, update_index)
+    return accum, lhs
+
+  accum = jnp.zeros((lhs.shape[0] * axis_size, rhs.shape[1]), dtype=lhs.dtype)
+  # fori_loop cause a crash: hlo_sharding.cc:817 Check failed: !IsManual()
+  # accum, lhs = jax.lax.fori_loop(0, axis_size - 1, f, (accum, lhs))
```

Comment:
Does this fori_loop still cause a crash?

Reply:
Oh good question... maybe @zhangqiaorjc can try it again? :D